xds/csds/csds.go

```diff
@@ -0,0 +1,321 @@
+			}
+		}
+		resources = append(resources, configDump)
+	}
+	return &v3statuspb.PerXdsConfig{
+		PerXdsConfig: &v3statuspb.PerXdsConfig_RouteConfig{
+			RouteConfig: &v3adminpb.RoutesConfigDump{
+				DynamicRouteConfigs: resources,
+			},
+		},
+	}
+}
+
+func (s *clientStatusServer) buildCDSPerXDSConfig() *v3statuspb.PerXdsConfig {
+	version, dump := s.xdsClient.DumpCDS()
+	var resources []*v3adminpb.ClustersConfigDump_DynamicCluster
+	for _, d := range dump {
+		configDump := &v3adminpb.ClustersConfigDump_DynamicCluster{
+			VersionInfo:  d.MD.Version,
+			ClientStatus: serviceStatusToProto(d.MD.Status),
+		}
+		if (d.MD.Timestamp != time.Time{}) {
+			configDump.Cluster = d.Raw
+			configDump.LastUpdated = timestamppb.New(d.MD.Timestamp)
+		}
+		if errState := d.MD.ErrState; errState != nil {
+			configDump.ErrorState = &v3adminpb.UpdateFailureState{
+				LastUpdateAttempt: timestamppb.New(errState.Timestamp),
+				Details:           errState.Err.Error(),
+				VersionInfo:       errState.Version,
+			}
+		}
+		resources = append(resources, configDump)
+	}
+	return &v3statuspb.PerXdsConfig{
+		PerXdsConfig: &v3statuspb.PerXdsConfig_ClusterConfig{
+			ClusterConfig: &v3adminpb.ClustersConfigDump{
+				VersionInfo:           version,
+				DynamicActiveClusters: resources,
+			},
+		},
+	}
+}
+
+func (s *clientStatusServer) buildEDSPerXDSConfig() *v3statuspb.PerXdsConfig {
+	_, dump := s.xdsClient.DumpEDS()
+	var resources []*v3adminpb.EndpointsConfigDump_DynamicEndpointConfig
+	for _, d := range dump {
+		configDump := &v3adminpb.EndpointsConfigDump_DynamicEndpointConfig{
+			VersionInfo:  d.MD.Version,
+			ClientStatus: serviceStatusToProto(d.MD.Status),
+		}
+		if (d.MD.Timestamp != time.Time{}) {
+			configDump.EndpointConfig = d.Raw
+			configDump.LastUpdated = timestamppb.New(d.MD.Timestamp)
+		}
+		if errState := d.MD.ErrState; errState != nil {
+			configDump.ErrorState = &v3adminpb.UpdateFailureState{
+				LastUpdateAttempt: timestamppb.New(errState.Timestamp),
+				Details:           errState.Err.Error(),
+				VersionInfo:       errState.Version,
+			}
+		}
+		resources = append(resources, configDump)
+	}
+	return &v3statuspb.PerXdsConfig{
+		PerXdsConfig: &v3statuspb.PerXdsConfig_EndpointConfig{
+			EndpointConfig: &v3adminpb.EndpointsConfigDump{
+				DynamicEndpointConfigs: resources,
+			},
+		},
+	}
+}
+
+func serviceStatusToProto(serviceStatus client.ServiceStatus) v3adminpb.ClientResourceStatus {
+	switch serviceStatus {
+	case client.ServiceStatusUnknown:
+		return v3adminpb.ClientResourceStatus_UNKNOWN
+	case client.ServiceStatusRequested:
+		return v3adminpb.ClientResourceStatus_REQUESTED
+	case client.ServiceStatusNotExist:
+		return v3adminpb.ClientResourceStatus_DOES_NOT_EXIST
+	case client.ServiceStatusACKed:
+		return v3adminpb.ClientResourceStatus_ACKED
+	case client.ServiceStatusNACKed:
+		return v3adminpb.ClientResourceStatus_NACKED
+	}
+	return v3adminpb.ClientResourceStatus_UNKNOWN
```

Comment:
Nit: move the last `return` into a `default` case?

Reply:
Done